pysph/cpy/translator.py

```diff
@@ -693,3 +703,57 @@ def __init__(self, detect_type=ocl_detect_type, known_types=None):
 
     def _get_self_type(self):
         return KnownType('GLOBAL_MEM %s*' % self._class_name)
+
+
+class CUDAConverter(OpenCLConverter):
+    def __init__(self, detect_type=ocl_detect_type, known_types=None):
+        super(CUDAConverter, self).__init__(detect_type, known_types)
+        self._local_decl = None
+
+    def _get_local_arg(self, arg, type):
+        return 'size_%s' % arg, 'int'
+
+    def _get_local_info(self, obj):
+        fname = obj.__name__
+        annotations = self._annotations[fname]
+        local_info = {}
+        for arg, kt in annotations.items():
+            if 'LOCAL_MEM' in kt.type:
+                local_info[arg] = kt.base_type
+        if local_info:
+            return local_info
+        return None
+
+    def parse_function(self, obj):
+        src = dedent(inspect.getsource(obj))
+        fname = obj.__name__
+        self._annotations[fname] = getattr(obj, '__annotations__', None)
+        self._local_decl = self._get_local_info(obj)
+        code = self.convert(src)
+        self._local_decl = None
+        self._annotations = {}
+        return code
+
+    def _get_local_declarations(self):
+        local_decl = ''
+        if self._local_decl:
+            decls = ['extern LOCAL_MEM float shared_buff[];']
+            # Reference:
+            # https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared
+            for arg, dtype in self._local_decl.items():
+                if len(decls) == 1:
+                    local_decl = ('%(dtype)s* %(arg)s = '
+                                  '(%(dtype)s*) shared_buff;')
+                    local_decl = local_decl % {'dtype': dtype, 'arg': arg}
+                    decls.append(local_decl)
+                    prev_arg = arg
+                else:
+                    local_decl = ('%(dtype)s* %(arg)s = (%(dtype)s*) '
+                                  '&%(prev_arg)s[size_%(prev_arg)s];')
```

Comment:
There is one big problem here if the sizes are not aligned to 4 bytes as discussed in the cuda programming guide.  So this should perhaps be carefully documented or perhaps a warning given if there is an alignment problem.